pkg/gpu/nvidia/health_check/health_checker.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/NVIDIA/gpu-monitoring-tools/bindings/go/nvml"
+	"github.com/golang/glog"
+
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+// GPUHealthChecker checks the health of nvidia GPUs.
+type GPUHealthChecker struct {
+	devices     map[string]pluginapi.Device
+	nvmlDevices map[string]*nvml.Device
+	health      chan pluginapi.Device
+	eventSet    nvml.EventSet
+	stop        chan bool
+}
+
+// NewGPUHealthChecker returns a GPUHealthChecker object for a given device name
+func NewGPUHealthChecker(devices map[string]pluginapi.Device, health chan pluginapi.Device) *GPUHealthChecker {
+	return &GPUHealthChecker{
+		devices:     devices,
+		nvmlDevices: make(map[string]*nvml.Device),
+		health:      health,
+		stop:        make(chan bool),
+	}
+}
+
+// Start registers NVML events and starts listening to them
+func (hc *GPUHealthChecker) Start() error {
+	glog.Info("Starting GPU Health Checker")
+
+	// Building mapping between device ID and their nvml represetation
+	count, err := nvml.GetDeviceCount()
+	if err != nil {
+		return fmt.Errorf("failed to get device count: %s", err)
+	}
+
+	glog.Infof("Found %d GPU devices", count)
+	for i := uint(0); i < count; i++ {
+		device, err := nvml.NewDeviceLite(i)
+		if err != nil {
+			return fmt.Errorf("failed to read device with index %d: %v", i, err)
+		}
+		deviceName, err := util.DeviceNameFromPath(device.Path)
+		if err != nil {
+			glog.Errorf("Invalid GPU device path found: %s. Skipping this device", device.Path)
+		}
+
+		if _, ok := hc.devices[deviceName]; !ok {
+			// we only monitor the devices passed in
+			glog.Warningf("Ignoring device %s for health check.", deviceName)
+			continue
+		}
+
+		glog.Infof("Found device %s for health monitoring. UUID: %s", deviceName, device.UUID)
+		hc.nvmlDevices[deviceName] = device
+	}
+
+	hc.eventSet = nvml.NewEventSet()
+	for _, d := range hc.nvmlDevices {
+		gpu, _, _, err := nvml.ParseMigDeviceUUID(d.UUID)
+		if err != nil {
+			gpu = d.UUID
+		}
+
+		glog.Infof("Registering device %v. UUID: %s", d.Path, d.UUID)
+		err = nvml.RegisterEventForDevice(hc.eventSet, nvml.XidCriticalError, gpu)
+		if err != nil {
+			if strings.HasSuffix(err.Error(), "Not Supported") {
+				glog.Warningf("Warning: %s is too old to support healthchecking: %v. It will always be marked healthy.", d.Path, err)
+				continue
+			} else {
+				return fmt.Errorf("failed to register device %s for NVML eventSet: %v", d.Path, err)
+			}
+		}
+	}
+
+	go func() {
+		if err := hc.listenToEvents(); err != nil {
+			glog.Errorf("GPUHealthChecker listenToEvents error: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+// chechHealth registers event with NVML to detect GPU critical errors
+func (hc *GPUHealthChecker) listenToEvents() error {
+	for {
+		select {
+		case <-hc.stop:
+			close(hc.stop)
+			return nil
+		default:
+		}
+
+		e, err := nvml.WaitForEvent(hc.eventSet, 5000)
+		if err != nil || e.Etype != nvml.XidCriticalError {
+			continue
```

Comment:
If `err != nil` should we print the error? maybe with lower verbosity.

Reply:
I didnt find the code exactly but i think this error most likely will always be the timeout errors. (WaitForEvent has a 5 second timeout)

Reply:
sounds good, in that case we can leave this as is.